### PR TITLE
Preserve prayer tracking in boost timer

### PR DESCRIPTION
**Pull request by Alexsuperfly**
From what i have been able to test there are 4 "sections" of 15 seconds each until your stats drain.
Preserve only needs to be on for the entirety of the last one for it to activate, once it does so you get 30 seconds extra even if you turn off preserve.

This will display the new decay timing while you have preserve on, if you don't turn it on soon enough or take it off too early it will catch that and display correctly.

This fulfills request #1059

### CODE REVIEW

**Comment by hijuiojiu** [comment]
The way it works is if you have preserve on for the last 15 seconds of the boost you will get 15 seconds of extra time. If you have it on for another 15 seconds after this you will get another 15 seconds of extra time. As in if you wish to get the full 30 seconds of boost you have to have preserve on for 30 seconds starting when the boost has 15 seconds left.

**Comment by Alexsuperfly** [comment]
not by what ive witnessed

File: runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java [hunk 1]
```diff
@@ -90,6 +91,10 @@
 
 	private BufferedImage overallIcon;
 
+	private boolean preserveBeenActive = false;
+
+	private int timeSinceChange;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java]: This should be a local variable

File: runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java [hunk 2]
```diff
@@ -191,8 +196,54 @@ public void addStatChangeIndicator()
...
+		if (timeSinceChange > 75)
+		{
+			return 90 - timeSinceChange;
+		}
+
+		preserveBeenActive = false;
+		return 60 - timeSinceChange;
 	}
+
+	public int timeSinceLastChange()
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java]: Make this private

File: runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java [hunk 3]
```diff
@@ -191,8 +194,54 @@ public void addStatChangeIndicator()
...
+	 * to gain full effect (seconds 45-75).
+	 *
+	 * @return integer value in seconds until next boost change
+	 */
 	public int getChangeTime()
 	{
-		return 60 - (int) Duration.between(lastChange, Instant.now()).getSeconds();
+		int timeSinceChange = timeSinceLastChange();
+
+		if (client.isPrayerActive(Prayer.PRESERVE) && timeSinceChange < 45)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java]: probably just store the state of preserve earlier and reuse it, no need for duplicated code.

Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java]: this block can probably just be merged with next if (e.g timeSinceChange < 45 || preserveBeenActive)

File: runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java [hunk 4]
```diff
@@ -191,8 +194,54 @@ public void addStatChangeIndicator()
...
+			return 90 - timeSinceChange;
+		}
+
+		if (timeSinceChange > 60 && timeSinceChange <= 75)
+		{
+			preserveBeenActive = false;
+			return 75 - timeSinceChange;
+		}
+
+		if (timeSinceChange > 75)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/boosts/BoostsPlugin.java]: This can probably be merged with first blocK?

**Comment by Alexsuperfly** [comment]
How does this look?

**Pull request merged**